src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker_Generators.cs

```diff
@@ -0,0 +1,159 @@
+            using var _2 = PooledHashSet<DocumentId>.GetInstance(out var allGeneratedDocumentIds);
+
+            var infos = infosOpt.Value;
+            foreach (var (documentIdentity, contentIdentity) in infos)
+            {
+                allGeneratedDocumentIds.Add(documentIdentity.DocumentId);
+
+                var existingDocument = generatorInfo.Documents.GetState(documentIdentity.DocumentId);
+                if (existingDocument?.Identity == documentIdentity)
+                {
+                    // ensure that the doc we have matches the content expected.
+                    if (existingDocument.GetTextChecksum() == contentIdentity.Checksum &&
+                        existingDocument.SourceText.Encoding?.WebName == contentIdentity.EncodingName &&
+                        existingDocument.SourceText.ChecksumAlgorithm == contentIdentity.ChecksumAlgorithm)
+                    {
+                        continue;
+                    }
+                }
+
+                // Couldn't find a matching generated doc.  Add this to the list to pull down.
+                documentsToAddOrUpdate.Add(documentIdentity.DocumentId);
+            }
+
+            // If we produced just as many documents as before, and none of them required any changes, then we can
+            // reuse the prior compilation.
+            if (infos.Length == generatorInfo.Documents.Count &&
+                documentsToAddOrUpdate.Count == 0 &&
+                compilationWithStaleGeneratedTrees != null)
+            {
+                return (compilationWithStaleGeneratedTrees, generatorInfo.WithDocumentsAreFinal(true));
+            }
+
+            // Either we generated a different number of files, and/or we had contents of files that changed. Ensure
+            // we know the contents of any new/changed files.
+            var generatedSourcesOpt = await connection.TryInvokeAsync(
+                solution,
+                this.ProjectState.Id,
+                (service, solutionChecksum, cancellationToken) => service.GetContentsAsync(
+                    solutionChecksum, projectId, documentsToAddOrUpdate.ToImmutable(), cancellationToken),
+                cancellationToken).ConfigureAwait(false);
+
+            if (!generatedSourcesOpt.HasValue)
+                return null;
+
+            var generatedSources = generatedSourcesOpt.Value;
+            Contract.ThrowIfTrue(generatedSources.Length != documentsToAddOrUpdate.Count);
+
+            // Now go through and produce the new document states, using what we have already if it is unchanged, or
+            // what we have retrieved for anything new/changed.
+            using var generatedDocumentsBuilder = TemporaryArray<SourceGeneratedDocumentState>.Empty;
+            foreach (var (documentIdentity, contentIdentity) in infos)
+            {
+                var addOrUpdateIndex = documentsToAddOrUpdate.IndexOf(documentIdentity.DocumentId);
```

Comment:
>var addOrUpdateIndex = documentsToAddOrUpdate.IndexOf(documentIdentity.DocumentId); [](http://example.com/codeflow?start=16&length=83)

Linear lookup is pretty gross. Are we pretty sure this won't be large?

Reply:
i'll add an index map :)

Reply:
done.

Reply:
A thing to behold....